### DISCUSSION TITLE
Fetch mandatory data before load App

### DISCUSSION THREAD

**Discussion by matteo-naif** [vercel/next.js#40582] [Help]
Hello, 
I need to fetch data (from an API endpoint) before the app starts (only the first time, no refetch on routing change).

I've read about the getInitialProps on _app file and I've tested it but it continue to fire.

Can you help me?

**Comment by icyJoseph** [comment]
I think the first time `getInitialProps` runs, it does so on the server, and then on SPA transitions, unless GSSP exists on the destination, it runs client side. You can check:

```ts
if(ctx.req) { /* fetch something */  return data};
return {};
```

The problem would be that when client side transitions happen, `getInitialProps` would still return something to your component, and you might want to ignore those "updates".

Does it help a bit?

## Edit

The above applies when middleware is not preset, if middleware exists, refer to https://github.com/vercel/next.js/issues/38918.

**Reply by matteo-naif** [reply]
I tested this code in a blank project and it runs on each page as well.

```js
MyApp.getInitialProps = async (appContext) => {
  // calls page's `getInitialProps` and fills `appProps.pageProps`
  
  if(appContext.ctx.req) {
  
    const appProps = await App.getInitialProps(appContext);
    const res = await fetch('https://jsonplaceholder.typicode.com/posts?_limit=9')
    const data = await res.json()

    console.log('Fetching data');

    return { ...appProps, data }
  
  };
  
  return {};

}
```

**Reply by icyJoseph** [reply]
Are you using middleware? Maybe you wanna take a look here: https://github.com/vercel/next.js/issues/38918

It is a bit tricky because, even after the fix for GSSP, it seems that GIP still runs every time.

**Reply by matteo-naif** [reply]
Unfortunately it didn't work and I don't use middleware. Everytime I load a page, the GIP from _app reload.
